### DISCUSSION TITLE
Is there AvaloniaEdit.TextMate package?

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [AvaloniaUI/AvaloniaEdit#179] [Q&A] [answered]
Hey. As far as I could understand, the demo is loading syntax packages from TextMate, but it doesn't seem to have a package. 

Also. Would you think it'd make sense to make syntax highlighting "native" for `TextEditor` type?

**Comment by danipen** [comment]
About the first part afaik @Takoooooo is working on it.

About the second question. What do you exactly mean by "make syntax highlight native"?

**Comment by WhiteBlackGoose** [comment]
@danipen what I mean is that the demo looks and works awesome, but it appears that setting the syntax highlighting is not so easy. I personally dream of just `syntaxHighlight="c#"` property, but I believe you didn't make it so to allow some configurability... maybe there's some trade off between convenience and configurability?

(note that I'm quite a noob in GUI in general, so maybe my comments aren't relevant)

**Comment by danipen** [comment]
Well, as you may know, AvaloniaEdit is a port of AvalonEdit. It comes with its own syntax highlight built-in engine, and this is the default one right now. We need to maintain it for backward compatibility.

We added TextMate as an extension, this is the reason it cannot work "built-in". We may change that in the future. 

But using TextMate is super easy. You can initialize text mate for a `TextEditor` in the following way:

```csharp
  TextEditor textEditor = new TextEditor();
  TextMate.TextMate.Installation textMate = textEditor.InstallTextMate(ThemeName.DarkPlus);
  textMate.SetGrammarByLanguageId("csharp");
```

**Reply by WhiteBlackGoose** [reply]
This looks easy, thanks. The reason I was asking is because I'm currently learning FuncUI, an MVU wrapper over Avalonia.FuncUI, for which it'd make sense to have it as a property. But I hope I'll be able to do it with attached properties... I'm [making](https://github.com/WhiteBlackGoose/MoreFuncUI#morefuncuiavaloniaedit) a FuncUI wrapper for your awesome component, so I guess I'll just use the TextMate extension via git modules.

**Reply by WhiteBlackGoose** [reply]
@danipen hey, sorry for stupid questions 😅 , but how can I unset grammar for a text editor?

**Comment by danipen** [comment]
Just set a null grammar for your `TextMate.Installation` object:
```csharp
textMate.SetGrammar(null);
```

**Reply by WhiteBlackGoose** [reply]
oh, so easy, thanks

**Reply by JaggerJo** [reply]
@WhiteBlackGoose with [this PR](https://github.com/fsprojects/Avalonia.FuncUI/pull/179#issuecomment-1006735097) it will be simpler to setup in FuncUI.

You will be able to obtain a reference to the underlaying component.

**Reply by JaggerJo** [reply]
This often makes wrapping components a lot simpler 

<img width="1040" alt="image" src="https://user-images.githubusercontent.com/13090415/148702441-6b709062-e316-4416-ae22-6684f6402ac0.png">

**Reply by WhiteBlackGoose** [reply]
Good to hear! I already [managed](https://github.com/WhiteBlackGoose/MoreFuncUI#morefuncuiavaloniaedit) to add some external properties to the component, now one can set the syntax highlighting with just one line. (though of course it makes it less capable than yours by default)

**Comment by Takoooooo** [comment, marked as answer]
Hey, we have recently [released](https://github.com/AvaloniaUI/AvaloniaEdit/releases/tag/0.10.12) this package, if you are still interested feel free to use :)

**Reply by JaggerJo** [reply]
Nice!

**Reply by WhiteBlackGoose** [reply]
👀 
Thanks, will give a look!

**Comment by SCLD-JBuczkowski** [comment]
I can't seem to get the example listed above to work.

            TextEditor textEditor = new TextEditor();
            TextMate.TextMate.Installation textMate = textEditor.InstallTextMate(ThemeName.DarkPlus);
            textMate.SetGrammarByLanguageId("csharp");

I was trying to use 1.0.56 (the latest release) and I was getting the compile errors.
The type 'IRegistryOptions' is defined in an assembly that is not referenced.
You must add a reference to assembly 'TextMateSharp, Version=1.0.29.0, Culture=neutral, PublicKeyToken=null'.

I tried dropping back to 1.0.29. It compiles and runs without crashing/exceptions.
However, it neither changes the theme nor appears to set the grammar correctly.
I see no syntax highlighting.

Has something changed in the latest releases which makes the existing example obsolete?

We would really like this functionality to be part of our release.

Thx
